### PR TITLE
Initial implementation of the libp2p daemon spec

### PR DESCRIPTION
**Pull request by jacobheun**
This implementation supports:

- **Connect**, connect the daemon to another peer
- **Identify**, get the PeerId and multiaddrs of the daemon
- **StreamOpen**, open a stream to a connected peer
- **StreamHandler**, register a stream handler for a set of protocols
- **ListPeers**, get a list of connected peers and their multiaddrs
- **DHT**, you can see the more details about what these do at https://github.com/libp2p/go-libp2p-daemon/blob/60e4ee83620ee34d7ebf29901fe1ec2e3d44ac63/specs/DHT.md
  - **FindPeer**
  - **FindProviders**
  - **Provide**
  - **GetClosestPeers**
  - **GetPublicKey**
  - **GetValue**
  - **PutValue**

### CODE REVIEW

File: src/daemon.js [hunk 1]
```diff
@@ -0,0 +1,460 @@
...
+            type: DHTResponse.Type.BEGIN
+          }
+        })]
+
+        const providers = await this.libp2p.contentRouting.findProviders(cid, {
+          maxNumProviders
+        })
+
+        providers.forEach(provider => {
+          responses.push(DHTResponse.encode({
```
[the remark this answers is not in the document]
Comment (jacobheun) [reply src/daemon.js]: This should be correct. Multiple responses may be written to the client for a dht request. The initial `OkResponse` signals that the request was valid and the included `DHTResponse` of type `BEGIN` signifies that the stream is starting. Then, the results will be streamed as individual `DHTResponse`s, until the final `DHTResponse` of type `END` is received. 

You can see this in the go implementation as well, for example https://github.com/libp2p/go-libp2p-daemon/blob/a08e63116baadbd95563064b5d72a71fc7046f24/dht.go#L175. The BEGIN response is added to the initial `OKResponse` and then the stream of DHTResponses, `rch`, is passed to the response stream.

Line 222 here is the needed, single, OKResponse.

[the remark this answers is not in the document]
Comment (vasco-santos) [reply src/daemon.js]: Makes sense, you are right! I changed that accordingly in the client!

**Comment by jacobheun** [comment]
@vasco-santos this is ready for a more thorough review.

File: src/libp2p.js [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+          return reject(err)
+        }
+        resolve(new PeerInfo(peerId))
+      })
+    }
+  })
+}
+
+class PeerRouting {
+  constructor (routing) {
```
Comment (vasco-santos) [inline src/libp2p.js]: we should add docs for the routing param

File: src/libp2p.js [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+    this._routing = routing
+  }
+
+  /**
+   *
+   * @param {PeerID} id The peer to find
+   * @param {object} options
+   * @param {number} options.maxTimeout How long the query should run
+   * @returns {Promise}
```
Comment (vasco-santos) [inline src/libp2p.js]: document what promise resolves to? There are some function here that document that, while others do not

(same below)

File: src/libp2p.js [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+  connMgrLo,
+  connMgrHi,
+  sock,
+  id
+} = {}) => {
+  const peerInfo = await getPeerInfo(id)
+  const peerBook = new PeerBook()
+  const bootstrapList = bootstrapPeers ? bootstrapPeers.split(',').filter(s => s !== '') : null
+
+  // TODO: Add multiaddrs
```
Comment (vasco-santos) [inline src/libp2p.js]: what is the plan here?

**Comment by vasco-santos** [comment]
Looks good @jacobheun . Left just small notes for you to consider

File: src/libp2p.js (hunk 4 above)
Comment (jacobheun) [reply src/libp2p.js]: I'll remove it, it's an old comment.

File: src/daemon.js [hunk 5]
```diff
@@ -0,0 +1,459 @@
...
+
+    let connection
+    let successfulProto
+    for (const protocol of proto) {
+      try {
+        connection = await this.libp2p.dial(peerInfo, protocol)
+        successfulProto = protocol
+        break
+      } catch (err) {
+        console.log(err)
```
[the remark this answers is not in the document]
Comment (jacobheun) [reply src/daemon.js]: yes we should! updated and switched to debug for non cli logging

**Comment by jacobheun** [comment]
Feedback should be resolved. I did a run through of the jsdocs, they should be filled out properly now.

**Comment by jacobheun** [comment]
@vasco-santos I bumped that dependency to a tag for now to avoid some volatility until a merge happens for that, or we decide to namespace a release

**Pull request merged**